src/unix.rs

```diff
@@ -567,3 +572,84 @@ pub fn lang() -> impl Iterator<Item = String> {
         index: Some(true),
     }
 }
+
+#[repr(C)]
+#[cfg(any(
+    target_os = "macos",
+    target_os = "ios",
+    target_os = "freebsd",
+    target_os = "dragonfly",
+    target_os = "netbsd",
+    target_os = "openbsd"
+))]
+struct UtsName {
+        #[cfg(not(target_os = "dragonfly"))]
+        pub sysname: [c_char; 256],
+        #[cfg(target_os = "dragonfly")]
+        pub sysname: [c_char; 32],
+        #[cfg(not(target_os = "dragonfly"))]
+        pub nodename: [c_char; 256],
+        #[cfg(target_os = "dragonfly")]
+        pub nodename: [c_char; 32],
+        #[cfg(not(target_os = "dragonfly"))]
+        pub release: [c_char; 256],
+        #[cfg(target_os = "dragonfly")]
+        pub release: [c_char; 32],
+        #[cfg(not(target_os = "dragonfly"))]
+        pub version: [c_char; 256],
+        #[cfg(target_os = "dragonfly")]
+        pub version: [c_char; 32],
+        #[cfg(not(target_os = "dragonfly"))]
+        pub machine: [c_char; 256],
+        #[cfg(target_os = "dragonfly")]
+        pub machine: [c_char; 32],
+}
+
+#[repr(C)]
+#[cfg(any(
+    target_os = "linux",
+    target_os = "android",
+    target_os = "fuchsia",
+    target_os = "redox"
+))]
+struct UtsName {
+    pub sysname: [c_char; 65],
+    pub nodename: [c_char; 65],
+    pub release: [c_char; 65],
+    pub version: [c_char; 65],
+    pub machine: [c_char; 65],
+    pub domainname: [c_char; 65],
+}
+
+// Buffer initialization
+impl Default for UtsName {
+    fn default() -> Self {
+        unsafe { mem::zeroed() }
+    }
+}
+
+extern "C" {
+    #[cfg(not(target_os = "freebsd"))]
+    fn uname(buf: *mut UtsName) -> c_int;
+
+    #[cfg(target_os = "freebsd")]
+    fn __xuname(nmln: c_int, buf: *mut c_void) -> c_int;
+}
+#[inline]
+#[cfg(target_os = "freebsd")]
+unsafe extern "C" fn uname(buf: *mut UtsName) -> c_int {
+     __xuname(256, buf as *mut c_void)
+}
+
+pub fn arch() -> Arch {
+    let mut buf = UtsName::default();
+    let result = unsafe { uname(&mut buf as *mut UtsName) };
```

Comment:
Should be able to use [`.cast()`](https://doc.rust-lang.org/std/primitive.pointer.html#method.cast) since it was added in 1.38